openvdb/openvdb/tools/Merge.h

```diff
@@ -0,0 +1,931 @@
+            if (insideTile.isOn(pos) || node.isChildMaskOn(pos))   {
+                node.addTile(pos, this->background(), iter.isValueOn());
+            }
+        }
+    }
+
+    // iterate over all children
+
+    for (auto iter = mergeNode->cbeginChildOn(); iter; ++iter) {
+        Index pos = iter.pos();
+        const Coord& ijk = iter.getCoord();
+        if (insideTile.isOn(pos)) {
+            auto childPtr = mTree.template stealOrDeepCopyNode<typename NodeT::ChildNodeType>(ijk);
+            if (childPtr) {
+                childPtr->resetBackground(this->otherBackground(), this->background());
+                childPtr->negate();
+                node.addChild(childPtr.release());
+            }
+        } else if (node.isChildMaskOn(pos)) {
+            // if both source and target are child nodes, merge recursion needs to continue
+            // along this branch to resolve the conflict
+            continueRecurse = true;
+        }
+    }
+
+    return continueRecurse;
+}
+
+template <typename TreeT>
+bool CsgDifferenceOp<TreeT>::operator()(LeafT& leaf, size_t) const
+{
+    using LeafT = typename TreeT::LeafNodeType;
+    using ValueT = typename LeafT::ValueType;
+    using BufferT = typename LeafT::Buffer;
+
+    // if buffer is not out-of-core and empty, leaf node must have only been
+    // partially constructed, so allocate and fill with background value
+
+    merge_internal::UnallocatedBuffer<BufferT, ValueT>::allocateAndFill(
+        leaf.buffer(), this->background());
+
+    const LeafT* mergeLeaf = mTree.template probeConstNode<LeafT>(leaf.origin());
+    if (!mergeLeaf)                 return false;
+
+    // if buffer is not out-of-core yet empty, leaf node must have only been
+    // partially constructed, so skip merge
+
+    if (merge_internal::UnallocatedBuffer<BufferT, ValueT>::isPartiallyConstructed(
+        mergeLeaf->buffer())) {
+        return false;
+    }
+
+    for (Index i = 0 ; i < LeafT::SIZE; i++) {
+        const ValueT& aValue = leaf.getValue(i);
+        ValueT bValue = math::negative(mergeLeaf->getValue(i));
+        if (aValue < bValue) { // a = max(a, -b)
+            leaf.setValueOnly(i, bValue);
+            leaf.setActiveState(i, mergeLeaf->isValueOn(i));
+        }
+    }
+
+    return false;
+}
+
+template <typename TreeT>
+const typename CsgDifferenceOp<TreeT>::ValueT&
+CsgDifferenceOp<TreeT>::background() const
+{
+    if (!mBackground) {
+        OPENVDB_THROW(RuntimeError, "Background value not set. "
+            "This operator is only intended to be used with foreachTopDown().");
+    }
+    return *mBackground;
+}
+
+template <typename TreeT>
+const typename CsgDifferenceOp<TreeT>::ValueT&
+CsgDifferenceOp<TreeT>::otherBackground() const
+{
+    if (!mOtherBackground) {
+        OPENVDB_THROW(RuntimeError, "Other background value not set. "
+            "This operator is only intended to be used with foreachTopDown().");
+    }
+    return *mOtherBackground;
```

Comment:
Can we remove these runtime errors in favour of asserts? The same at line 721. nounwind/noexcept  opts are always nice (though I realise some other internal methods might throw, we can still avoid unnecessary ones - these are private after all).

Reply:
I agree, however, I note that we should have non-crashing recovery code for mOtherBackground that makes sense.

(I'm also reminded I want us to use a VDB_ASSERT for better control as well...)

Reply:
I'll take a look at this, thanks.